subsys/net/lib/wifi_credentials/wifi_credentials.c

```diff
@@ -0,0 +1,538 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+#include <zephyr/init.h>
+#include <net/wifi_credentials.h>
+#include <sys/types.h>
+#include <stdlib.h>
+
+LOG_MODULE_REGISTER(wifi_credentials, CONFIG_WIFI_CREDENTIALS_LOG_LEVEL);
+
+#define ENTRY_MAX_LEN sizeof(struct wifi_credentials_personal)
+
+
+static K_MUTEX_DEFINE(wifi_credentials_mutex);
+
+static char configured_ssids[CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES][WIFI_SSID_MAX_LEN];
+
+/**
+ * @brief Stores settings entry in flash
+ *
+ * @param idx credential index
+ * @param buf encoded settings entry
+ * @param buf_len length of buf
+ * @return 0 on success, otherwise a negative error code
+ */
+int wifi_credentials_serialize_entry(size_t idx, const void *buf, size_t buf_len);
+
+/**
+ * @brief Deletes settings entry from flash
+ *
+ * @param idx credential index
+ * @return 0 on success, otherwise a negative error code
+ */
+int wifi_credentials_delete_entry(size_t idx);
+
+/**
+ * @brief Loads settings entry from flash
+ *
+ * @param idx credential index
+ * @param buf encoded settings entry
+ * @param buf_len length of buf
+ * @return 0 on success, otherwise a negative error code
+ */
+int wifi_credentials_deserialize_entry(size_t idx, void *buf, size_t buf_len);
+
+/**
+ * @brief Finds index of given SSID if it exists.
+ *
+ * @param ssid		SSID to look for (buffer of WIFI_SSID_MAX_LEN length)
+ * @return		index if entry is found, -1 otherwise
+ */
+static inline ssize_t lookup_idx(const uint8_t *ssid)
+{
+	for (size_t i = 0; i < CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES; ++i) {
+		if (strcmp(ssid, configured_ssids[i]) == 0) {
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+/**
+ * @brief Finds unused index to store new entry at.
+ *
+ * @return		index if empty slot is found, -1 otherwise
+ */
+static inline ssize_t lookup_unused_idx(void)
+{
+	for (size_t i = 0; i < CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES; ++i) {
+		if (strlen(configured_ssids[i]) == 0) {
```

Comment:
Although not enforced by the standard, most implementations expect SSIDs to be null-terminated. Android seems to support 32 Byte long SSIDs (without termination), but you cannot type in null characters anyways. I think I'll implement the check as follows: If all characters in the SSID are NULL, the entry is considered unused.

Reply:
A theoretical caveat of this implementation is as follows: SSIDs that are just NULL characters are not supported. This is not explicitly checked in the implementation, but will result in an error code trying to load non-existent indices.
Also, this implementation cannot distinguish between SSIDs that have different amounts of NULL characters on the end.

I think these scenarios are not worth the effort.

Reply:
I agree with your conclusion, all-NULL SSIDs are not worth the effort to support, but I am not an expert on these things.

Though I think if you are going to treat all-NULL SSIDs as not valid, you should directly check for them and raise an error when configuring them, especially since configuration should hopefully be a fairly rare event. You could write a macro or inline function to do the all-null check.

Reply:
(When starting to integrate this into the provisioning sample, I actually changed my mind. Now I save SSID lengths explicitly.)